### PR TITLE
pipeline for cache ova and vagrant from a pre-build image

### PR DESCRIPTION
**Pull request by PengTian0**
Pipeline for cache ova and vagrant.
1. Add a pipeline: PackerCacheBuild/Jenkinsfile
    This pipeline will build a template VMX and OVF which OS installed and all RackHD dependency installed.
2. Change the pipelines : build_ova.groovy and build_vagrant.groovy.
     The 2 pipelines support build image from the above VMX and OVF.

The PR depends on: https://github.com/RackHD/RackHD/pull/704

### CODE REVIEW

**Comment by PengTian0** [comment]
@panpan0000 @changev @anhou @iceiilin

File: jobs/build_ova/build_ova.sh [hunk 1]
```diff
@@ -1,24 +1,34 @@
-#!/bin/bash
+#!/bin/bash -x
+
 set +e
-#sudo cp ${HOME}/bin/packer   /usr/bin
-#sudo apt-get install -y  jq
+mv $WORKSPACE/cache_image/packer/* $WORKSPACE/build/packer/
```
Comment (anhou) [inline jobs/build_ova/build_ova.sh]: I noticed there are *.md5 file under packer/ directory, Is it feasible to check the md5sum before using existing images. just a thought and confirm with you, or maybe it's not essential?

**Review by anhou — APPROVED** [review]

File: jobs/build_ova/build_ova.sh (hunk 1 above)
Comment (PengTian0) [reply jobs/build_ova/build_ova.sh]: The "*.md5" is the md5 file of ova which is generated at the last step.

File: jobs/PackerCacheBuild/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                        set -e
+                        ./HWIMO-BUILD
+                        popd
+                        '''
+                        archiveArtifacts "RackHD/packer/output-*/*.*"
+                    } finally{
+                        sh '''#!/bin/bash -x
+                        set +e
+                        pkill packer
+                        rm ~/VirtualBox\\ VMs/rackhd-${OS_VER} -rf
```
Comment (panpan0000) [inline jobs/PackerCacheBuild/Jenkinsfile]: Seems we need to add  ```rm  -r ~/workspace/PACKER_CACHE_BUILD/packer/output-virtualbox-iso/rackhd-${OS_VER}.vdi``` in cleanup

File: jobs/PackerCacheBuild/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    checkout scm
+    shareMethod = load("jobs/ShareMethod.groovy")
+}
+
+def cacheBuild(ArrayList<String> used_resources, String BUILD_TYPE){
+    String label_name = "packer"
+    lock(label:label_name,quantity:1){
+        String node_name = shareMethod.occupyAvailableLockedResource(label_name, used_resources)
+        try{
+            node(node_name){
```
Comment (panpan0000) [inline jobs/PackerCacheBuild/Jenkinsfile:27]: Do you think we need timeout ?

**Review by panpan0000 — APPROVED** [review]
Thank you very much, @Tian !
Let's prove more the test results

File: jobs/PackerCacheBuild/Jenkinsfile (hunk 3 above)
Comment (PengTian0) [reply jobs/PackerCacheBuild/Jenkinsfile:27]: Thanks!
I added timeout for ```sh``` step.

Comment (PengTian0) [inline jobs/PackerCacheBuild/Jenkinsfile]: Yes!
I added timeout for ```sh``` step.

**Comment by panpan0000** [comment]
Sandbox Jenkins job demo:
http://xx.xx.59.175:8080/job/PACKER_CACHE_BUILD1/
http://xx.xx.59.175:8080/job/PACKER_BUILD_WITH_CACHE/
Demo Jenkins job: 
http://rackhdci.lss.emc.com/job/Archiving/job/PACKER_CACHE_BUILD/
http://rackhdci.lss.emc.com/job/RAC4360MasterCI/

File: jobs/PackerCacheBuild/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import groovy.transform.Field;
```
Comment (panpan0000) [inline jobs/PackerCacheBuild/Jenkinsfile:1]: [!!!!!]
I recalled another more important thing, there're 2 triggers  of the "cache-build" job:
(1) any SCM change in RackHD/packer folder
```
checkout(
[
  $class: 'GitSCM',
  branches: [[name: '*/master']],
  doGenerateSubmoduleConfigurations: false,
  extensions: [[$class: 'PathRestriction', excludedRegions: '', includedRegions: 'packer/(.*/)?.*']],
  submoduleCfg: [], userRemoteConfigs: [[url: 'https://github.com/RackHD/RackHD/']]
]
)
```
(2) We need to specific the "Build Trigger" to "daily" in the Jenkins job configuration as well.

Comment (PengTian0) [reply jobs/PackerCacheBuild/Jenkinsfile:1]: Thanks!
Fixed.

**Pull request merged**